File: src/insights/tools.ts
// De værktøjer insight-agenten må bruge.
//
// Princip: værktøjerne regner, modellen vurderer og formulerer. Alle tal i et
// udkast skal kunne findes i et værktøjsresultat (se verify.ts), så de
// afledte tal (gennemsnit, rang, procentvis ændring) beregnes her og ikke af
// modellen.
//
// Værktøjerne er rene funktioner over et `Dataset` (alle kampe i ligaen
// indlæst én gang pr. kørsel), så de kan testes uden database.

import type Anthropic from "@anthropic-ai/sdk";

export type MatchRow = {
  id: string;
  season: string; // "2025/2026"
  round: number | null;
  kickoff: Date;
  home: string; // holdnavn
  homeSlug: string;
  away: string;
  awaySlug: string;
  homeScore: number | null;
  awayScore: number | null;
  attendance: number;
  capacity: number | null;
  note: string | null;
};

export type RecentDraft = {
  id: string;
  createdAt: Date;
  status: string;
  kind: string;
  headline: string;
  reviewNote: string | null;
};

export type Dataset = {
  league: string;
  matches: MatchRow[]; // kun kampe med tilskuertal, sorteret efter kickoff
  newMatchIds: string[];
  recentDrafts: RecentDraft[];
};

// ------------------------------------------------------------------ helpers

const mean = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
const avg = (xs: number[]) => (xs.length ? Math.round(mean(xs)) : null);
const round1 = (x: number) => Math.round(x * 10) / 10;
const pct = (now: number | null, before: number | null) =>
  now != null && before ? round1(((now - before) / before) * 100) : null;

function median(xs: number[]): number {
  const s = [...xs].sort((a, b) => a - b);
  const m = Math.floor(s.length / 2);
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}

function stdev(xs: number[]): number | null {
  if (xs.length < 3) return null;
  const m = mean(xs);
  return Math.sqrt(xs.reduce((a, x) => a + (x - m) ** 2, 0) / (xs.length - 1));
}

export function previousSeason(season: string): string {
  const [a, b] = season.split("/").map(Number);
  return b ? `${a - 1}/${b - 1}` : String(a - 1);
}

const dkTime = new Intl.DateTimeFormat("da-DK", {
  timeZone: "Europe/Copenhagen",
  year: "numeric", month: "2-digit", day: "2-digit", hour: "2-digit", minute: "2-digit",
});

/** Dansk tid som "19.07.2025 18.00", så modellen ikke skal omregne fra UTC. */
export const formatKickoff = (d: Date) => dkTime.format(d).replace(",", "");

function matchInfo(m: MatchRow) {
  return {
    match_id: m.id,
    season: m.season,
    round: m.round,
    kickoff: formatKickoff(m.kickoff),
    home: m.home,
    away: m.away,
    score: m.homeScore != null && m.awayScore != null ? `${m.homeScore}-${m.awayScore}` : null,
    attendance: m.attendance,
    stadium_capacity: m.capacity,
    pct_of_capacity: m.capacity ? round1((m.attendance / m.capacity) * 100) : null,
    // Admin-note om særlige forhold, fx en lukket tribune - vigtig kontekst!
    note: m.note,
  };
}

function findMatch(ds: Dataset, id: string): MatchRow {
  const m = ds.matches.find((x) => x.id === id);
  if (!m) throw new Error(`Ukendt match_id ${id}`);
  return m;
}

function findTeamSlug(ds: Dataset, team: string): string {
  const t = team.toLowerCase();
  const m = ds.matches.find((x) => x.homeSlug === t || x.home.toLowerCase() === t);
  if (!m) throw new Error(`Ukendt hold "${team}". Brug slug fra list_available_data.`);
  return m.homeSlug;
}

// -------------------------------------------------------------------- tools

export function listNewMatches(ds: Dataset) {
  const ids = new Set(ds.newMatchIds);
  const rows = ds.matches.filter((m) => ids.has(m.id));
  return { league: ds.league, count: rows.length, matches: rows.map(matchInfo) };
}

export function listAvailableData(ds: Dataset) {
  const seasons = new Map<string, MatchRow[]>();
  const teams = new Map<string, string>();
  for (const m of ds.matches) {
    seasons.set(m.season, [...(seasons.get(m.season) ?? []), m]);
    teams.set(m.homeSlug, m.home);
  }
  return {
    league: ds.league,
    seasons: [...seasons].map(([season, ms]) => ({
      season,
      matches: ms.length,
      last_round: Math.max(0, ...ms.map((m) => m.round ?? 0)),
      total_attendance: ms.reduce((a, m) => a + m.attendance, 0),
    })),
    teams: [...teams].map(([slug, name]) => ({ slug, name })),
  };
}

const SCOPES = {
  league_season: "alle ligakampe i samme sæson",
  league_all: "alle ligakampe i databasen (alle sæsoner)",
  home_team_season: "hjemmeholdets hjemmekampe i samme sæson",
  home_team_all: "hjemmeholdets hjemmekampe i alle sæsoner",
  fixture_history: "samme opgør (samme hjemme- og udehold) på tværs af sæsoner",
} as const;
type Scope = keyof typeof SCOPES;

/**
 * Placerer en kamps tilskuertal i en sammenligningsgruppe. Kun kampe spillet
 * til og med denne kamps kickoff tæller med, så "rekord" betyder rekord på
 * tidspunktet - og svaret ændrer sig ikke, hvis agenten kører igen senere.
 */
export function compareAttendance(ds: Dataset, input: { match_id: string; scope: string }) {
  const scope = input.scope as Scope;
  if (!(scope in SCOPES)) throw new Error(`scope skal være en af ${Object.keys(SCOPES).join(", ")}`);
  const m = findMatch(ds, input.match_id);
  const inScope: Record<Scope, (x: MatchRow) => boolean> = {
    league_season: (x) => x.season === m.season,
    league_all: () => true,
    home_team_season: (x) => x.homeSlug === m.homeSlug && x.season === m.season,
    home_team_all: (x) => x.homeSlug === m.homeSlug,
    fixture_history: (x) => x.homeSlug === m.homeSlug && x.awaySlug === m.awaySlug,
  };
  const group = ds.matches.filter((x) => inScope[scope](x) && x.kickoff <= m.kickoff);
  const values = group.map((x) => x.attendance);
  const earlier = group.filter((x) => x.id !== m.id);
  const byAtt = [...earlier].sort((a, b) => b.attendance - a.attendance);
  const prevHigh = byAtt[0];
  const prevLow = byAtt[byAtt.length - 1];
  const mu = mean(values);
  const sd = stdev(values);

  return {
    match: matchInfo(m),
    scope,
    comparison_group: SCOPES[scope],
    matches_in_group: values.length,
    rank_highest_first: 1 + values.filter((v) => v > m.attendance).length,
    mean_attendance: Math.round(mu),
    median_attendance: Math.round(median(values)),
    diff_vs_mean: Math.round(m.attendance - mu),
    pct_vs_mean: pct(m.attendance, mu),
    z_score: sd ? Math.round(((m.attendance - mu) / sd) * 100) / 100 : null,
    is_highest_so_far: !prevHigh || m.attendance > prevHigh.attendance,
    is_lowest_so_far: !prevLow || m.attendance < prevLow.attendance,
    previous_highest: prevHigh ? matchInfo(prevHigh) : null,
    previous_lowest: prevLow ? matchInfo(prevLow) : null,
    warning: values.length < 5 ? "Grupper med under 5 kampe siger meget lidt." : null,
  };
}

/** Et holds hjemmekampe i en sæson, udviklingen i de seneste kampe og
 *  sammenligning med samme antal hjemmekampe sidste sæson. */
export function teamAttendanceTrend(ds: Dataset, input: { team: string; season: string; window?: number }) {
  const slug = findTeamSlug(ds, input.team);
  const window = input.window ?? 3;
  const home = ds.matches.filter((m) => m.homeSlug === slug && m.season === input.season);
  if (!home.length) throw new Error(`Ingen hjemmekampe for ${input.team} i ${input.season}`);
  const values = home.map((m) => m.attendance);
  const n = values.length;
  const last = values.slice(-window);
  const before = values.slice(-2 * window, -window);
  const hasBefore = before.length === window;

  const prevLabel = previousSeason(input.season);
  const prev = ds.matches
    .filter((m) => m.homeSlug === slug && m.season === prevLabel)
    .slice(0, n)
    .map((m) => m.attendance);
  const samePoint = prev.length === n;

  return {
    team: home[0].home,
    season: input.season,
    home_matches: home.map((m) => ({
      match_id: m.id, kickoff: formatKickoff(m.kickoff), opponent: m.away, attendance: m.attendance, note: m.note,
    })),
    home_matches_played: n,
    season_avg: avg(values),
    window,
    avg_last_window: avg(last),
    avg_previous_window: hasBefore ? avg(before) : null,
    pct_change_last_vs_previous_window: hasBefore ? pct(avg(last), avg(before)) : null,
    previous_season: prevLabel,
    previous_season_avg_same_point: samePoint ? avg(prev) : null,
    pct_vs_previous_season_same_point: samePoint ? pct(avg(values), avg(prev)) : null,
    warning: samePoint ? null : `Ingen sammenlignelige data for ${prevLabel} (fx op-/nedrykning).`,
  };
}

/** Ligaen til og med en runde, sammenlignet med sidste sæson til og med samme runde. */
export function seasonOverview(ds: Dataset, input: { season: string; up_to_round?: number }) {
  const inSeason = (s: string) => ds.matches.filter((m) => m.season === s && m.round != null);
  const cur = inSeason(input.season);
  if (!cur.length) throw new Error(`Ingen kampe med runde i ${input.season}`);
  const upTo = input.up_to_round ?? Math.max(...cur.map((m) => m.round!));

  const summary = (rows: MatchRow[]) => {
    const r = rows.filter((m) => m.round! <= upTo);
    if (!r.length) return null;
    const byTeam = new Map<string, number[]>();
    for (const m of r) byTeam.set(m.home, [...(byTeam.get(m.home) ?? []), m.attendance]);
    return {
      matches: r.length,
      total_attendance: r.reduce((a, m) => a + m.attendance, 0),
      avg_attendance: avg(r.map((m) => m.attendance)),
      home_avg_by_team: Object.fromEntries(
        [...byTeam].map(([t, v]) => [t, avg(v)!] as const).sort((a, b) => b[1] - a[1])
      ),
    };
  };

  const current = summary(cur)!;
  const prevLabel = previousSeason(input.season);
  const previous = summary(inSeason(prevLabel));
  return {
    season: input.season,
    up_to_round: upTo,
    current,
    previous_season: prevLabel,
    previous_same_rounds: previous,
    pct_change_avg_attendance: previous ? pct(current.avg_attendance, previous.avg_attendance) : null,
    pct_change_total_attendance: previous ? pct(current.total_attendance, previous.total_attendance) : null,
  };
}

export function roundSummary(ds: Dataset, input: { season: string; round_number: number }) {
  const season = ds.matches.filter((m) => m.season === input.season && m.round != null);
  const rows = season.filter((m) => m.round === input.round_number).sort((a, b) => b.attendance - a.attendance);
  if (!rows.length) throw new Error(`Ingen kampe i runde ${input.round_number} af ${input.season}`);
  const byRound = new Map<number, number[]>();
  for (const m of season) byRound.set(m.round!, [...(byRound.get(m.round!) ?? []), m.attendance]);
  const thisAvg = mean(rows.map((m) => m.attendance));
  return {
    season: input.season,
    round: input.round_number,
    matches: rows.map(matchInfo),
    total_attendance: rows.reduce((a, m) => a + m.attendance, 0),
    avg_attendance: Math.round(thisAvg),
    rounds_played_in_season: byRound.size,
    rank_of_round_by_avg: 1 + [...byRound.values()].filter((v) => mean(v) > thisAvg).length,
  };
}

export function listRecentDrafts(ds: Dataset) {
  return {
    drafts: ds.recentDrafts.map((d) => ({
      created: formatKickoff(d.createdAt), status: d.status, kind: d.kind,
      headline: d.headline, review_note: d.reviewNote,
    })),
  };
}

// ------------------------------------------------------------------ schemas
// Beskrivelserne er modellens eneste dokumentation af værktøjerne - de er en
// del af prompten og værd at skrive omhyggeligt.

const noInput = { type: "object" as const, properties: {}, additionalProperties: false };

export const TOOL_SCHEMAS: Anthropic.Beta.BetaTool[] = [
  {
    name: "list_new_matches",
    description: "De kampe, der er kommet til siden sidste analyse. Start her.",
    input_schema: noInput,
  },
  {
    name: "list_available_data",
    description: "Sæsoner i databasen (antal kampe, seneste runde, samlet tilskuertal) og holdenes slug + navn.",
    input_schema: noInput,
  },
  {
    name: "compare_attendance",
    description:
      "Placerer én kamps tilskuertal i en sammenligningsgruppe: rang, gennemsnit, median, procentvis " +
      "afvigelse, z-score, udnyttelse af stadionkapacitet, og om det er det højeste/laveste hidtil (med den " +
      "tidligere rekord). Kun kampe til og med kampens kickoff tæller. Brug flere scopes for at se, om noget " +
      "er usædvanligt for ligaen, for holdet eller for netop dette opgør.",
    input_schema: {
      type: "object",
      properties: {
        match_id: { type: "string" },
        scope: { type: "string", enum: Object.keys(SCOPES) },
      },
      required: ["match_id", "scope"],
      additionalProperties: false,
    },
  },
  {
    name: "team_attendance_trend",
    description:
      "Et holds hjemmekampe i en sæson med gennemsnit, udvikling i de seneste `window` kampe mod de " +
      "`window` før, og sammenligning med samme antal hjemmekampe sidste sæson.",
    input_schema: {
      type: "object",
      properties: {
        team: { type: "string", description: "Holdets slug, fx 'fc-kobenhavn'" },
        season: { type: "string", description: "Fx '2026/2027'" },
        window: { type: "integer", minimum: 1, maximum: 10 },
      },
      required: ["team", "season"],
      additionalProperties: false,
    },
  },
  {
    name: "season_overview",
    description:
      "Ligaens samlede og gennemsnitlige tilskuertal til og med en runde, pr. hold, og sammenlignet med " +
      "sidste sæson til og med samme runde.",
    input_schema: {
      type: "object",
      properties: {
        season: { type: "string" },
        up_to_round: { type: "integer", description: "Udelad for seneste spillede runde" },
      },
      required: ["season"],
      additionalProperties: false,
    },
  },
  {
    name: "round_summary",
    description: "Alle kampe i én runde med samlet og gennemsnitligt tilskuertal, og rundens placering blandt sæsonens runder.",
    input_schema: {
      type: "object",
      properties: { season: { type: "string" }, round_number: { type: "integer" } },
      required: ["season", "round_number"],
      additionalProperties: false,
    },
  },
  {
    name: "list_recent_drafts",
    description:
      "De seneste insight-udkast med status og redaktørens note. Tjek dem, så du ikke gentager et insight, " +
      "og lær af begrundelserne for afviste udkast.",
    input_schema: noInput,
  },
  {
    name: "submit_draft",
    description:
      "Gem et insight-udkast til godkendelse. Hvert tal i headline og body skal stå i et af dine " +
      "værktøjsresultater, ellers afvises udkastet med en fejl, du kan rette. evidence skal pege på " +
      "\"ref\"-feltet (fx \"R3\") i de værktøjssvar, tallene kommer fra.",
    input_schema: {
      type: "object",
      properties: {
        kind: { type: "string", enum: ["record", "outlier", "trend", "milestone", "other"] },
        headline: { type: "string", description: "Kort overskrift, max ca. 80 tegn" },
        body: { type: "string", description: "1-3 sætninger på dansk" },
        match_ids: { type: "array", items: { type: "string" } },
        evidence: {
          type: "array",
          minItems: 1,
          items: {
            type: "object",
            properties: {
              claim: { type: "string", description: "Påstanden, beviset understøtter" },
              ref: { type: "string", description: "\"ref\"-feltet fra værktøjssvaret, fx \"R3\"" },
            },
            required: ["claim", "ref"],
            additionalProperties: false,
          },
        },
      },
      required: ["kind", "headline", "body", "match_ids", "evidence"],
      additionalProperties: false,
    },
  },
];

// submit_draft håndteres i agent-loopet, fordi det skal kende kørslens
// tidligere værktøjsresultater. Resten er rene dataværktøjer.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const DATA_TOOLS: Record<string, (ds: Dataset, input: any) => unknown> = {
  list_new_matches: listNewMatches,
  list_available_data: listAvailableData,
  compare_attendance: compareAttendance,
  team_attendance_trend: teamAttendanceTrend,
  season_overview: seasonOverview,
  round_summary: roundSummary,
  list_recent_drafts: listRecentDrafts,
};

File: tilskuerstats/tools.py
"""De værktøjer agenten må bruge.

Princip: værktøjerne regner, modellen vurderer og formulerer. Alle tal i et
udkast skal kunne findes i et værktøjsresultat (se verify.py), så de
afledte tal (gennemsnit, rang, procentvis ændring) beregnes her og ikke af
modellen.

Hvert værktøj er en almindelig Python-funktion `fn(conn, **input) -> dict`
plus et JSON-schema, som sendes til Claude i `tools`-parameteren.
"""
from __future__ import annotations

import json
import sqlite3
import statistics
from typing import Any, Callable

# Kun klubber vi er sikre på. Ukendte forkortelser vises bare som koden.
TEAM_NAMES = {
    "AGF": "AGF", "AaB": "AaB", "BIF": "Brøndby IF", "FCK": "FC København",
    "FCM": "FC Midtjylland", "FCN": "FC Nordsjælland", "LBK": "Lyngby Boldklub",
    "OB": "OB", "RFC": "Randers FC", "SIF": "Silkeborg IF", "SJF": "Sønderjyske",
    "VB": "Vejle Boldklub", "VFF": "Viborg FF", "FCF": "FC Fredericia",
    "EFB": "Esbjerg fB", "ACH": "AC Horsens",
}


# ---------------------------------------------------------------- helpers

def _match_dict(row: sqlite3.Row) -> dict:
    return {
        "match_id": row["id"],
        "season": row["season"],
        "round": row["round"],
        "kickoff": row["kickoff"],
        "home": row["home"],
        "away": row["away"],
        "score": f"{row['home_goals']}-{row['away_goals']}",
        "attendance": row["attendance"],
    }


def _get_match(conn: sqlite3.Connection, match_id: int) -> sqlite3.Row:
    row = conn.execute("SELECT * FROM matches WHERE id = ?", (match_id,)).fetchone()
    if row is None:
        raise ValueError(f"Ukendt match_id {match_id}")
    return row


def _pct(new: float, old: float) -> float | None:
    return round((new - old) / old * 100, 1) if old else None


def _avg(values: list[int]) -> int | None:
    return round(statistics.mean(values)) if values else None


def _previous_season(season: str) -> str:
    a, b = (int(y) for y in season.split("/"))
    return f"{a - 1}/{b - 1}"


# ------------------------------------------------------------------ tools

def list_new_matches(conn: sqlite3.Connection) -> dict:
    rows = conn.execute(
        "SELECT * FROM matches WHERE analyzed_at IS NULL ORDER BY kickoff LIMIT 60"
    ).fetchall()
    return {"count": len(rows), "matches": [_match_dict(r) for r in rows]}


def list_available_data(conn: sqlite3.Connection) -> dict:
    seasons = conn.execute(
        """SELECT season, COUNT(*) AS matches, MAX(round) AS last_round,
                  SUM(attendance) AS total_attendance
           FROM matches GROUP BY season ORDER BY season"""
    ).fetchall()
    teams = [r["home"] for r in conn.execute("SELECT DISTINCT home FROM matches ORDER BY home")]
    return {
        "seasons": [dict(s) for s in seasons],
        "teams": [{"code": t, "name": TEAM_NAMES.get(t, t)} for t in teams],
    }


COMPARE_SCOPES = {
    "league_season": ("season = :season", "alle ligakampe i samme sæson"),
    "league_all": ("1 = 1", "alle ligakampe i databasen (alle sæsoner)"),
    "home_team_season": ("home = :home AND season = :season", "hjemmeholdets hjemmekampe i samme sæson"),
    "home_team_all": ("home = :home", "hjemmeholdets hjemmekampe i alle sæsoner"),
    "fixture_history": ("home = :home AND away = :away", "samme opgør (samme hjemme- og udehold) på tværs af sæsoner"),
}


def compare_attendance(conn: sqlite3.Connection, match_id: int, scope: str) -> dict:
    """Placerer en kamps tilskuertal i en sammenligningsgruppe.

    Kun kampe spillet til og med denne kamps kickoff tæller med, så et
    'rekord'-resultat betyder rekord på tidspunktet - og svaret ændrer sig
    ikke, hvis agenten kører igen senere.
    """
    if scope not in COMPARE_SCOPES:
        raise ValueError(f"scope skal være en af {sorted(COMPARE_SCOPES)}")
    m = _get_match(conn, match_id)
    where, description = COMPARE_SCOPES[scope]
    params = {"season": m["season"], "home": m["home"], "away": m["away"], "kickoff": m["kickoff"]}
    rows = conn.execute(
        f"SELECT * FROM matches WHERE {where} AND kickoff <= :kickoff ORDER BY kickoff", params
    ).fetchall()

    values = [r["attendance"] for r in rows]
    att = m["attendance"]
    earlier = [r for r in rows if r["id"] != m["id"]]
    prev_high = max(earlier, key=lambda r: r["attendance"], default=None)
    prev_low = min(earlier, key=lambda r: r["attendance"], default=None)
    mean = statistics.mean(values)
    stdev = statistics.stdev(values) if len(values) >= 3 else None

    return {
        "match": _match_dict(m),
        "scope": scope,
        "comparison_group": description,
        "matches_in_group": len(values),
        "rank_highest_first": 1 + sum(v > att for v in values),
        "mean_attendance": round(mean),
        "median_attendance": round(statistics.median(values)),
        "diff_vs_mean": round(att - mean),
        "pct_vs_mean": _pct(att, mean),
        "z_score": round((att - mean) / stdev, 2) if stdev else None,
        "is_highest_so_far": prev_high is None or att > prev_high["attendance"],
        "is_lowest_so_far": prev_low is None or att < prev_low["attendance"],
        "previous_highest": _match_dict(prev_high) if prev_high else None,
        "previous_lowest": _match_dict(prev_low) if prev_low else None,
        "note": "Grupper med under 5 kampe siger meget lidt." if len(values) < 5 else None,
    }


def team_attendance_trend(conn: sqlite3.Connection, team: str, season: str, window: int = 3) -> dict:
    """Hjemmekampe for et hold i en sæson + udvikling og sammenligning med
    samme tidspunkt sidste sæson (samme antal hjemmekampe)."""
    rows = conn.execute(
        "SELECT * FROM matches WHERE home = ? AND season = ? ORDER BY kickoff", (team, season)
    ).fetchall()
    if not rows:
        raise ValueError(f"Ingen hjemmekampe for {team} i {season}")
    values = [r["attendance"] for r in rows]
    n = len(values)

    last, before = values[-window:], values[-2 * window:-window]
    prev_season = _previous_season(season)
    prev_rows = conn.execute(
        "SELECT attendance FROM matches WHERE home = ? AND season = ? ORDER BY kickoff LIMIT ?",
        (team, prev_season, n),
    ).fetchall()
    prev_values = [r["attendance"] for r in prev_rows]
    same_point = len(prev_values) == n

    return {
        "team": team,
        "team_name": TEAM_NAMES.get(team, team),
        "season": season,
        "home_matches": [
            {"match_id": r["id"], "kickoff": r["kickoff"], "opponent": r["away"],
             "attendance": r["attendance"]} for r in rows
        ],
        "home_matches_played": n,
        "season_avg": _avg(values),
        f"avg_last_{window}": _avg(last),
        f"avg_previous_{window}": _avg(before) if len(before) == window else None,
        "pct_change_last_vs_previous": _pct(_avg(last), _avg(before)) if len(before) == window else None,
        "previous_season": prev_season,
        "previous_season_avg_same_point": _avg(prev_values) if same_point else None,
        "pct_vs_previous_season_same_point": _pct(_avg(values), _avg(prev_values)) if same_point else None,
        "note": None if same_point else f"Ingen sammenlignelige data for {prev_season} (fx op-/nedrykning).",
    }


def season_overview(conn: sqlite3.Connection, season: str, up_to_round: int | None = None) -> dict:
    """Ligaens samlede tilskuertal til og med en runde, sammenlignet med
    sidste sæson til og med samme runde (like-for-like)."""
    if up_to_round is None:
        up_to_round = conn.execute(
            "SELECT MAX(round) FROM matches WHERE season = ?", (season,)
        ).fetchone()[0]
        if up_to_round is None:
            raise ValueError(f"Ingen kampe i {season}")

    def summary(s: str) -> dict | None:
        rows = conn.execute(
            "SELECT * FROM matches WHERE season = ? AND round <= ?", (s, up_to_round)
        ).fetchall()
        if not rows:
            return None
        by_team: dict[str, list[int]] = {}
        for r in rows:
            by_team.setdefault(r["home"], []).append(r["attendance"])
        return {
            "matches": len(rows),
            "total_attendance": sum(r["attendance"] for r in rows),
            "avg_attendance": _avg([r["attendance"] for r in rows]),
            "home_avg_by_team": dict(sorted(
                ((t, _avg(v)) for t, v in by_team.items()), key=lambda kv: -kv[1]
            )),
        }

    cur = summary(season)
    if cur is None:
        raise ValueError(f"Ingen kampe i {season}")
    prev_season = _previous_season(season)
    prev = summary(prev_season)
    return {
        "season": season,
        "up_to_round": up_to_round,
        "current": cur,
        "previous_season": prev_season,
        "previous_same_rounds": prev,
        "pct_change_avg_attendance": _pct(cur["avg_attendance"], prev["avg_attendance"]) if prev else None,
        "pct_change_total_attendance": _pct(cur["total_attendance"], prev["total_attendance"]) if prev else None,
    }


def round_summary(conn: sqlite3.Connection, season: str, round_number: int) -> dict:
    rows = conn.execute(
        "SELECT * FROM matches WHERE season = ? AND round = ? ORDER BY attendance DESC",
        (season, round_number),
    ).fetchall()
    if not rows:
        raise ValueError(f"Ingen kampe i runde {round_number} af {season}")
    per_round = conn.execute(
        "SELECT round, AVG(attendance) AS avg FROM matches WHERE season = ? GROUP BY round",
        (season,),
    ).fetchall()
    this_avg = statistics.mean(r["attendance"] for r in rows)
    return {
        "season": season,
        "round": round_number,
        "matches": [_match_dict(r) for r in rows],
        "total_attendance": sum(r["attendance"] for r in rows),
        "avg_attendance": round(this_avg),
        "rounds_played_in_season": len(per_round),
        "rank_of_round_by_avg": 1 + sum(r["avg"] > this_avg for r in per_round),
    }


def list_recent_drafts(conn: sqlite3.Connection, limit: int = 15) -> dict:
    rows = conn.execute(
        "SELECT id, created_at, status, kind, headline, review_note FROM drafts ORDER BY id DESC LIMIT ?",
        (limit,),
    ).fetchall()
    return {"drafts": [dict(r) for r in rows]}


# ---------------------------------------------------------------- schemas
# Beskrivelserne er modellens eneste dokumentation af værktøjerne - de er
# en del af prompten og værd at skrive omhyggeligt.

TOOL_SCHEMAS: list[dict[str, Any]] = [
    {
        "name": "list_new_matches",
        "description": "Returnerer de kampe, der er kommet til siden sidste analyse. Start her.",
        "input_schema": {"type": "object", "properties": {}, "additionalProperties": False},
    },
    {
        "name": "list_available_data",
        "description": "Oversigt over sæsoner i databasen (antal kampe, seneste runde, samlet tilskuertal) og holdkoder med klubnavne.",
        "input_schema": {"type": "object", "properties": {}, "additionalProperties": False},
    },
    {
        "name": "compare_attendance",
        "description": (
            "Placerer én kamps tilskuertal i en sammenligningsgruppe: rang, gennemsnit, median, "
            "procentvis afvigelse, z-score og om det er den højeste/laveste hidtil (med den tidligere "
            "rekord). Kun kampe til og med kampens kickoff tæller med. Brug flere scopes for at se, "
            "om noget er usædvanligt for ligaen, for holdet eller for netop dette opgør."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "match_id": {"type": "integer"},
                "scope": {"type": "string", "enum": sorted(COMPARE_SCOPES)},
            },
            "required": ["match_id", "scope"],
            "additionalProperties": False,
        },
    },
    {
        "name": "team_attendance_trend",
        "description": (
            "Et holds hjemmekampe i en sæson med gennemsnit, udvikling i de seneste `window` kampe "
            "mod de `window` før, og sammenligning med samme antal hjemmekampe sidste sæson."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "team": {"type": "string", "description": "Holdkode, fx 'FCK'"},
                "season": {"type": "string", "description": "Fx '2025/2026'"},
                "window": {"type": "integer", "minimum": 1, "maximum": 10, "default": 3},
            },
            "required": ["team", "season"],
            "additionalProperties": False,
        },
    },
    {
        "name": "season_overview",
        "description": (
            "Ligaens samlede og gennemsnitlige tilskuertal til og med en runde, pr. hold, og "
            "sammenlignet med sidste sæson til og med samme runde."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "season": {"type": "string"},
                "up_to_round": {"type": "integer", "description": "Udelad for seneste spillede runde"},
            },
            "required": ["season"],
            "additionalProperties": False,
        },
    },
    {
        "name": "round_summary",
        "description": "Alle kampe i én runde med samlet og gennemsnitligt tilskuertal, og rundens placering blandt sæsonens runder.",
        "input_schema": {
            "type": "object",
            "properties": {"season": {"type": "string"}, "round_number": {"type": "integer"}},
            "required": ["season", "round_number"],
            "additionalProperties": False,
        },
    },
    {
        "name": "list_recent_drafts",
        "description": "De seneste insight-udkast med status og redaktørens note (også godkendte/afviste). Tjek dem, så du ikke gentager et insight, og lær af begrundelserne for afviste udkast.",
        "input_schema": {
            "type": "object",
            "properties": {"limit": {"type": "integer", "minimum": 1, "maximum": 50}},
            "additionalProperties": False,
        },
    },
    {
        "name": "submit_draft",
        "description": (
            "Gem et insight-udkast til menneskelig godkendelse. Hvert tal i headline og body skal "
            "stå i et af dine værktøjsresultater, ellers afvises udkastet med en fejl, du kan rette. "
            "evidence skal pege på de tool_use-id'er, som tallene kommer fra."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "kind": {"type": "string", "enum": ["record", "outlier", "trend", "milestone", "other"]},
                "headline": {"type": "string", "description": "Kort overskrift, max ca. 80 tegn"},
                "body": {"type": "string", "description": "1-3 sætninger på dansk"},
                "match_ids": {"type": "array", "items": {"type": "integer"}},
                "evidence": {
                    "type": "array",
                    "minItems": 1,
                    "items": {
                        "type": "object",
                        "properties": {
                            "claim": {"type": "string", "description": "Den påstand, beviset understøtter"},
                            "tool_use_id": {"type": "string", "description": "id på det tool_use-kald, der viser det"},
                        },
                        "required": ["claim", "tool_use_id"],
                        "additionalProperties": False,
                    },
                },
            },
            "required": ["kind", "headline", "body", "match_ids", "evidence"],
            "additionalProperties": False,
        },
    },
]

# submit_draft håndteres af agent-loopet, fordi det skal kende kørslens
# tidligere værktøjsresultater. Resten er rene dataværktøjer.
DATA_TOOLS: dict[str, Callable[..., dict]] = {
    "list_new_matches": list_new_matches,
    "list_available_data": list_available_data,
    "compare_attendance": compare_attendance,
    "team_attendance_trend": team_attendance_trend,
    "season_overview": season_overview,
    "round_summary": round_summary,
    "list_recent_drafts": list_recent_drafts,
}


def run_data_tool(conn: sqlite3.Connection, name: str, tool_input: dict) -> dict:
    if name not in DATA_TOOLS:
        raise ValueError(f"Ukendt værktøj: {name}")
    return DATA_TOOLS[name](conn, **tool_input)


def to_json(result: Any) -> str:
    return json.dumps(result, ensure_ascii=False)
